### PR TITLE
[2.067.0-b2][regression] fix Issues 14212 and 6586

### PR DESCRIPTION
**Pull request by 9il**
ping @MartinNowak

### CODE REVIEW

**Comment by 9il** [comment]
Oh, ignore first commit.

File: std/math.d [hunk 1]
```diff
@@ -2178,9 +2178,10 @@ creal expi(real y) @trusted pure nothrow @nogc
  *      $(TR $(TD $(PLUSMN)$(NAN)) $(TD $(PLUSMN)$(NAN)) $(TD int.min))
  *      )
  */
-T frexp(T)(T value, out int exp) @trusted pure nothrow @nogc
+T frexp(T)(T _value, out int exp) @trusted pure nothrow @nogc
```
Comment (kuettler) [inline std/math.d]: Would it be ok to not use names start with underscores as function parameters? This comes up in the public documentation.

Comment (9il) [inline std/math.d]: Thanks! Fixed

**Comment by WalterBright** [comment]
I don't understand - what is this attempting to fix?

**Comment by 9il** [comment]
Reduced test case:
```D
int exp;
const double val; // const or immutable data
auto l = frexp(val, exp); // fails until this PR
```

File: std/math.d [hunk 2]
```diff
@@ -2181,11 +2181,12 @@ creal expi(real y) @trusted pure nothrow @nogc
 T frexp(T)(T value, out int exp) @trusted pure nothrow @nogc
     if(isFloatingPoint!T)
 {
-    ushort* vu = cast(ushort*)&value;
+    Unqual!T vf = value;
```
Comment (9il) [inline std/math.d]: Mutable `vf`

File: std/math.d [hunk 3]
```diff
@@ -2379,7 +2380,7 @@ unittest
 {
     import std.typetuple, std.typecons;
 
-    foreach (T; TypeTuple!(real, double, float))
+    foreach (T; TypeTuple!(const(real), const(double), const(float)))
```
Comment (MartinNowak) [inline std/math.d]: You should keep the old test.

Comment (9il) [inline std/math.d]: fixed

File: std/math.d (hunk 2 above)
Comment (MartinNowak) [inline std/math.d]: What does `vf` stand for, can't make any sense out of it.

**Comment by MartinNowak** [comment]
Only the subnormal case requires a change to the argument. Might be better to simply add another `return value * F.RECIP_EPSILON` there.

File: std/math.d (hunk 2 above)
Comment (9il) [inline std/math.d]: The `value` can be const or immutable. Should it be `in`: `(in T value, out int exp)`?

**Comment by 9il** [comment]
No
```D
-            value *= F.RECIP_EPSILON;
+            vf *= F.RECIP_EPSILON;
             ex = vu[F.EXPPOS_SHORT] & F.EXPMASK; /// vu pointer to vf/value
             exp = ((ex - F.EXPBIAS) >> 7) - T.mant_dig + 1; /// out data
```

File: std/math.d (hunk 2 above)
Comment (MartinNowak) [inline std/math.d]: I meant the name `vf`.

Yes, using `in` or `const` makes sense here, because it reduces the number of template instantiations.

Comment (9il) [inline std/math.d]: OK, will add `const` instead new name.

**Comment by MartinNowak** [comment]
Ah, forgot about the aliasing of `vu`.

File: std/math.d [hunk 4]
```diff
@@ -2409,13 +2410,13 @@ unittest
 
         static if (floatTraits!(T).realFormat == RealFormat.ieeeExtended)
         {
-            static T[3][] extendedvals = [ // x,frexp,exp
+            static T[3][] extendedvals2 = [ // x,frexp,exp
```
Comment (MartinNowak) [inline std/math.d]: Why did you change the name?

Comment (MartinNowak) [inline std/math.d]: Ah, because now real occurs multiple times.

Comment (MartinNowak) [inline std/math.d]: Simply move that test out of the loop and directly use `real`.

**Comment by 9il** [comment]
fixed

**Comment by MartinNowak** [comment]
Auto-merge toggled on

**Comment by MartinNowak** [comment]
thx
